File: 07_linked_list/07_02_reverse_linked_list_recursion/reverse_linked_list_recursion.js
class ListNode {
    constructor(val = 0, next = null) {
        this.val = val;
        this.next = next;
    }
}

function reverseList(head) {
    if (head == null || head.next == null) {
        return head;
    }
    const reversed = reverseList(head.next);
    head.next.next = head;
    head.next = null;
    return reversed;
}

// Example usage
const head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
process.stdout.write("Before reverse: ");
let trav = head;
while (trav !== null) {
    process.stdout.write(trav.val + " ");
    trav = trav.next;
}
console.log();

const reversedHead = reverseList(head);

// Print the reversed list for verification
process.stdout.write("After reverse: ");
let reversedTrav = reversedHead;
while (reversedTrav !== null) {
    process.stdout.write(reversedTrav.val + " ");
    reversedTrav = reversedTrav.next;
}
console.log();

File: 07_linked_list/07_02_reverse_linked_list_recursion/ReverseLinkedListRecursion.java
public class ReverseLinkedListRecursion {
    public static void main(String[] args) {
        // Example usage
        ListNode head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
        System.out.print("Before reverse: ");
        ListNode trav = head;
        while (trav != null) {
            System.out.print(trav.val + " ");
            trav = trav.next;
        }
        System.out.println();
        ListNode reversedHead = reverseList(head);
        // Print the reversed list for verification
        System.out.print("After reverse: ");
        while (reversedHead != null) {
            System.out.print(reversedHead.val + " ");
            reversedHead = reversedHead.next;
        }
    }

    public static ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode reversed = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return reversed;
    }
}

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}